Search by keyword in Manage Stock grid

giacmir opened the pull request:
fixes #1222

<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios,
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description
When searching for a keyword in Manage Stock grid, the search filteron name is considered a fulltext search and then translated in a 'equal to' query for collection filtering.
I have added a StockCollectionProcessor virtual type that uses a StockFilterProcessor to apply a StockNameFilter, that eventually translates the fulltext filter to a like query.

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. magento-engcom/msi#1222: Cannot search stocks in stock grid by keyword #1222 
2. magento-engcom/msi#325: Search by keyword by part of the name of Stock doesn't work in Manage Stocks grid #325

### Manual testing scenarios
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. New Stock "Test Stock 1" created
2. Try to do "Search by keyword" with word "Test"
3. Test Stock 1 is returned
4. Try to add further filters to search, as ID
5. Correct results are returned

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

giacmir committed:
Search by keyword in stock grid

fixes #1222

naydav commented on app/code/Magento/Inventory/Model/Stock/SearchCriteria/CollectionProcessor/StockFilterProcessor/StockNameFilter.php:
```diff
@@ -0,0 +1,39 @@
+
+        if ($conditionType === 'fulltext') {
+            $conditionType = 'like';
+            $value = '%' . $value . '%';
```
Looks like we need to quote input value
Also, need to consider if the value will contain symbols like `%` or `_`

giacmir replied:
For quoting are you saying that I should use the mysql quoting function like `$collection->getConnection()->quote()`? I can't find any other reference in the codebase of a 'like' condition being quoted. Can you better explain this change request?

naydav commented on app/code/Magento/Inventory/etc/di.xml:
```diff
@@ -69,6 +69,28 @@
+        <arguments>
+            <argument name="processors" xsi:type="array">
+                <item name="filters" xsi:type="object">Magento\Inventory\Model\Stock\SearchCriteria\CollectionProcessor\StockFilterProcessor</item>
+                <item name="sorting" xsi:type="object">Magento\Framework\Api\SearchCriteria\CollectionProcessor\SortingProcessor</item>
```
Looks like we don't need to specify sorting and pagination (equals to default value)

naydav commented on app/code/Magento/Inventory/Model/Stock/SearchCriteria/CollectionProcessor/StockFilterProcessor/StockNameFilter.php:
```diff
@@ -0,0 +1,39 @@
+     * @param Filter $filter
+     * @param AbstractDb $collection
+     * @return bool Whether the filter was applied
+     * @since 100.2.0
```
Extra annotation

naydav commented on app/code/Magento/Inventory/Model/Stock/SearchCriteria/CollectionProcessor/StockFilterProcessor/StockNameFilter.php:
```diff
@@ -0,0 +1,39 @@
+<?php
+/**
```
Missed declare strict type

naydav commented on app/code/Magento/Inventory/Model/Stock/SearchCriteria/CollectionProcessor/StockFilterProcessor/StockNameFilter.php:
```diff
@@ -0,0 +1,39 @@
+use Magento\Framework\Api\SearchCriteria\CollectionProcessor\FilterProcessor\CustomFilterInterface;
+use Magento\Framework\Data\Collection\AbstractDb;
+
+class StockNameFilter implements CustomFilterInterface
```
I'm not sure that we need to bind this filter only for name...
You can reuse this filter for another fields... or another entities, for exampe we have the same problem in source grid

My propouse is using something like `FieldSearchFilter` for class name

giacmir replied:
This was meant to be used only for this specific grid. For a more generic solution won't be better something like #1285 that operates at framework level?

naydav commented on app/code/Magento/Inventory/Model/Stock/SearchCriteria/CollectionProcessor/StockFilterProcessor/StockNameFilter.php:
```diff
@@ -0,0 +1,39 @@
+     * @return bool Whether the filter was applied
+     * @since 100.2.0
+     */
+    public function apply(Filter $filter, AbstractDb $collection)
```
Use type hinting for return values

naydav commented on app/code/Magento/Inventory/Model/Stock/SearchCriteria/CollectionProcessor/StockFilterProcessor/StockNameFilter.php:
```diff
@@ -0,0 +1,39 @@
+            $conditionType = 'like';
+            $value = '%' . $value . '%';
+        }
+        $nameFilter = [$conditionType => [$value]];
```
It will be useful not only for name (I mean need to change variable name)

giacmir committed:
Requested changes for #1343

- removed extra arguments for filter processor
- fixed extra annotations and missing strict type declaration
- added type hinting

naydav commented:
Closed due to https://github.com/magento-engcom/msi/pull/1285
